### PR TITLE
change dataInput.Matchers -> dataOutput.Conditions

### PR DESCRIPTION
**Pull request by wonderflow**
Signed-off-by: 天元 <jianbo.sjb@alibaba-inc.com>

### CODE REVIEW

File: apis/core/v1alpha2/core_types.go [hunk 1]
```diff
@@ -384,6 +386,12 @@ type DataOutput struct {
 
 	// FieldPath refers to the value of an object's field.
 	FieldPath string `json:"fieldPath,omitempty"`
+
+	// Matchers specify the requirements to match a value.
+	// Different requirements are AND-ed together.
+	// If no matcher is specified, it is by default to check not empty.
+	// +optional
+	Matchers []DataMatcherRequirement `json:"matchers,omitempty"`
```
Comment (hongchaodeng) [inline apis/core/v1alpha2/core_types.go:394]: If we put this into DataOutput, this doesn't mean "Matchers" anymore.
It is "Conditions":

```
conditions:
- fieldPath: status.state
  op: eq
  value: Running
```

Comment (hongchaodeng) [reply apis/core/v1alpha2/core_types.go:394]: And the comment needs to be changed:

```Go
// Conditions specify the conditions that should be satisfied before emitting a data output.
Conditions []ConditionRequirement
```

**Comment by hongchaodeng** [comment]
The idea and code LGTM.
I think the name should change a bit.

**Comment by hongchaodeng** [comment]
lgtm

File: examples/dependency/README.md [hunk 2]
```diff
@@ -1,15 +1,173 @@
...
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: source
+      dataOutputs:
+        - name: example-key
+          fieldPath: "spec.key"
+          matchers:
```
Comment (zzxwill) [inline examples/dependency/README.md:112]:
```suggestion
          conditions:
```

Comment (zzxwill) [reply examples/dependency/README.md:112]: Please also replace `matchers` in files and filenames.

File: apis/core/v1alpha2/core_types.go (hunk 1 above)
Comment (zzxwill) [reply apis/core/v1alpha2/core_types.go]: Putting `Conditions` in `dataOutputs` parts seems confusing to me. If the condition turns out to be `false`, I might mistakenly think the outer controller will prevent the deployment of the component. 

Besides, if we set two conditions C1 and C2 for component A, and component B and C depends on component A, there will be confusion when B and C can move on.
By keeping the original design, ie C1 and C2 set in component B and C respectively, if C1 == true, component B can go on; if C2 == false, component C will hold on. This situation seems more efficient. But I admit I could not make a real example for this situation.

File: examples/dependency/README.md [hunk 3]
```diff
@@ -1,15 +1,173 @@
...
+```shell script
+$ kubectl get foo.example.com source -o yaml
+apiVersion: example.com/v1
+kind: Foo
+metadata:
+  name: source
+  namespace: default
+```
+
+And the workload of `sink` component also don't have `spec.key`:
```
Comment (ryanzhang-oss) [inline examples/dependency/README.md:52]: the sink should not exist!

File: pkg/dependency/dag.go [hunk 4]
```diff
@@ -28,6 +28,8 @@ type SinksPerSource struct {
 type Source struct {
 	// ObjectRef refers to the object this source come from.
 	ObjectRef *corev1.ObjectReference
+
+	Matchers []v1alpha2.ConditionRequirement
```
Comment (ryanzhang-oss) [inline pkg/dependency/dag.go:32]: Would you consider not using ObjectReference and just spell out all the needed fields?  Not sure why we need to borrow an object reference

File: examples/dependency/README.md (hunk 3 above)
Comment (wonderflow) [reply examples/dependency/README.md:52]: yes, I'll check it

File: pkg/dependency/dag.go (hunk 4 above)
Comment (wonderflow) [reply pkg/dependency/dag.go:32]: Why don't you want borrow object reference here? They're exactly same

**Pull request merged**